lib/connection/transaction.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Janus.Connection.Transaction do
+  @moduledoc false
+
+  require Logger
+  @spec init_transaction_call_table() :: :ets.tab()
+  def init_transaction_call_table() do
+    :ets.new(:pending_calls, [:duplicate_bag, :private])
+  end
+
+  # Generates a transaction ID for the payload and ensures that it is unused
+  @spec generate_transaction!(:ets.tab()) :: binary
+  def generate_transaction!(pending_calls_table) do
+    transaction = :crypto.strong_rand_bytes(32) |> Base.encode64()
+
+    case :ets.lookup(pending_calls_table, transaction) do
+      [] ->
+        transaction
+
+      _ ->
+        generate_transaction!(pending_calls_table)
+    end
+  end
+
+  @spec insert_transaction(:ets.tab(), binary, GenServer.from(), integer) :: true
+  def insert_transaction(pending_calls_table, transaction, from, timeout) do
```

Comment:
Separating transaction generation and insertion is problematic as it creates a race condition - you may generate 2 the same ids before the first is put in ETS. Lookup and insertion should be atomic.